### DISCUSSION TITLE
Enhancing Taqueria's Compilation Process for Consistency and Ease-of-use

### DISCUSSION THREAD

**Discussion by mweichert** [ecadlabs/taqueria#1894] [Ideas]
# Compilation using LIGO and SmartPy

This discussion will focus on the compilation process for the LIGO and SmartPy compiler, although we should also consider other compilers like Archetype.

## Understanding the Compilers

In the Tezos ecosystem, LIGO and SmartPy are two popular languages for writing smart contracts. However, they have different behaviors when it comes to compiling smart contracts from source files:

- **LIGO**: The LIGO compiler requires the `-m` option to specify a single module to compile from a source file. If a source file contains multiple modules, the compiler needs to be invoked separately for each module. There is no concept of a default module to compile. It's important to note that a module in LIGO doesn't necessarily have to be a smart contract. It could be a library, utility, or a general organizational mechanism. A smart contract in LIGO is identified by a module that has a function with an `@entry` annotation or a `main` function. If a source file doesn't contain a module but contains an `@entry` or `main` function, then that entire file is treated as a smart contract.

- **SmartPy**: The SmartPy compiler looks for a scenario in the source file, and within the scenario, one or more smart contracts can be instantiated. For each instantiation, a separate smart contract is created. This means that a single source file can yield multiple compiled smart contracts. A smart contract in SmartPy is defined as a subclass of `sp.SmartContract`.

## Current Implementation in Taqueria

In Taqueria, our development tool for Tezos, the current implementation of the `compile` task assumes that a single source file contains one smart contract. This works well for LIGO, but not for SmartPy where multiple contracts can be defined in a single file.

Additionally, our `compile-all` task for the LIGO plugin uses custom scanning logic to find and compile all modules in each source file. For SmartPy, we need a similar approach to compile all smart contracts defined in a source file.

## Proposed Solution

We propose enhancing Taqueria's compilation process with the following changes:

1. **Add Contract/Module Parameter**: We'll introduce a new parameter `--contract [contractName]` to the `compile` task. This parameter will be required for LIGO and optional for SmartPy. For LIGO, it will specify the module to compile. For SmartPy, if the `--contract` parameter is not provided, Taqueria will compile all contracts defined in the source file. Note, I think to remain consistent with LIGO terminology, we'll alias `--contract` to `--module`.

2. **Enhance `compile-all` Task**: For LIGO, we'll adjust our `compile-all` task to use LIGO's `list-declarations` command instead of our custom scanning logic. This should provide a more robust and accurate way to find and compile all modules. For SmartPy, we'll search for all subclasses of `sp.SmartContract` in each source file, generate a new SmartPy file that creates a scenario, and instantiate each contract for compilation.

## Organizing the Output

To make the output more organized, we'll create a subdirectory in the `artifacts` folder for each smart contract. This will make it easier to find the compiled output for each contract, especially in larger projects with many contracts. While this isn't directly related to the core problem of handling multiple contracts in a single source file, we believe it's an important aspect of improving the overall user experience with Taqueria.

We believe these changes will make Taqueria more flexible and user-friendly, especially for developers working with SmartPy. We welcome any feedback or suggestions on this proposed solution.

**Comment by Laucans** [comment]
More context about list-declaration, 
given a contract : 
```jsligo

namespace ContractWithMain {
  type storage = int;
  type parameter = int;
  type ret = [list<operation>, storage];
  // Three entrypoints

  const main = ([p, s]: [parameter, storage]): ret => [list([]), 1]
};

namespace ContractWithAnnotation {
  type storage = int;
  type parameter = int;
  type ret = [list<operation>, storage];
  // Three entrypoints

  @entry
  const entrypoint = ([p, s]: [parameter, storage]): ret => [list([]), 1]
};

namespace NotAContract {
  let int = 1
};

```

The `list-declarations`command will : 
```shell
ligo info list-declaration contracts/pokeGame.jsligo --display-format json
{
  "source_file": "contracts/pokeGame.jsligo",
  "declarations": [
    "NotAContract.int", "ContractWithAnnotation.$contract",
    "ContractWithAnnotation.$views", "ContractWithAnnotation.$main",
    "ContractWithAnnotation.entrypoint", "ContractWithMain.main"
  ]
}
```

So you can identify contract with the ModuleName.($main|main)

**Reply by Laucans** [reply]
But it still possible to have a contract without this. Example
```    
    type storage = int;
    type parameter = int;
    type ret = [list<operation>, storage];
    // Three entrypoints

    const notMainName = ([p, s]: [parameter, storage]): ret => [list([]), 1]

```

Function notMainName respect the signature of a main function and can be compiled with:
```
ligo compile contract contracts/pokeGame.jsligo -e notMainName
```

In this case, I'm not sure we are able to determine it automatically without explicit declaration somewhere...

**Reply by mweichert** [reply]
Thanks for the thoughts and details, @Laucans. The above looks like an edge case that might not be too common. Do you think it would be ok if `compile-all` missed the above, but add an option to LIGO's `compile` task that would allow you to compile that contract on its own using something like this: `taq compile pokeGame.jsligo --entry nonMainName`?

**Reply by Laucans** [reply]
I agree, keep the edge case documented and we are good :)
